castervoice/rules/core/keyboard_rules/keyboard.py

```diff
@@ -0,0 +1,147 @@
+from castervoice.lib.merge.state.actions2 import UntilCancelled
+from castervoice.lib.merge.state.short import S, L, R
+
+_tpd = text_punc_dict()
+
+class Keyboard(MergeRule):
+    pronunciation = "keyboard"
+    
+    mapping = {
+        "<modifier> <button_dictionary_1>":
+              R(Key("%(modifier)s%(button_dictionary_1)s"),
+              rdescript="press modifiers plus buttons from button_dictionary_1, non-repeatable"),
+        "<hold_release> [<modifier>] <button_dictionary_1>":
+              R(Key("%(modifier)s%(button_dictionary_1)s:%(hold_release)s"),
+              rdescript="press modifiers plus buttons from button_dictionary_1, non-repeatable"),
+    }
+
+    # These buttons can be used without using the "press" prefix.
+    right_spec = "(right | ross)"
+    left_spec = "(left | lease)"
+    button_dictionary_1 = {
+        "(F{}".format(i) + " | function {})".format(i) : "f{}".format(i)
+        for i in range(1, 13)
+    }
+    button_dictionary_1.update(caster_alphabet())
+    button_dictionary_1.update(_tpd)
+    longhand_punctuation_names = {
+        "minus": "hyphen",
+        "hyphen": "hyphen",
+        "comma": "comma",
+        "deckle": "colon",
+        "colon": "colon",
+        "slash": "slash",
+        "backslash": "backslash"
+    }
+    button_dictionary_1.update(longhand_punctuation_names)
+    shift_spec = "(shift | shin)"
+    control_spec = "(control | fly)"
+    alt_spec = "alt"
+    windows_spec = "windows"
+    # in the punctuation dictionary it uses " " which is not the correct dragonfly key name.
+    del button_dictionary_1["ace"]
+    button_dictionary_1.update({
+        "(tab | tabby)": "tab",
+        "(backspace | clear)": "backspace",
+        "(delete | deli)": "del",
+        "(enter | shock)": "enter",
+        left_spec: "left",
+        right_spec: "right",
+        "(up | sauce)": "up",
+        "(down | dunce)": "down",
+        "page (down | dunce)": "pgdown",
+        "page (up | sauce)": "pgup",
+        "(ace | space)": "space",
+        "zero": "0",
+        "one": "1",
+        "two": "2",
+        "three": "3",
+        "four": "4",
+        "five": "5",
+        "six": "6",
+        "seven": "7",
+        "eight": "8",
+        "nine": "9",
+        shift_spec: "shift",
+        control_spec: "control",
+        alt_spec: "alt",
+        right_spec + shift_spec: "rshift",
+        right_spec + control_spec: "rcontrol",
+        right_spec + alt_spec: "ralt",
+        SymbolSpecs.CANCEL: "escape",
+        "insert": "insert",
+        "pause": "pause",
+        windows_spec: "win",
+        "(apps | popup)": "apps",
+        "print screen": "printscreen",
+        "scroll lock": "scrolllock",
+        "num lock": "numlock",
+        "caps lock": "capslock",
+        "(home | lease wally | latch)": "home",
+        "(end | ross wally | ratch)": "end",
+        # number pad numbers deliberately left off
```

Comment:
I definitely agree with these choices here. We should continue to elevate those things into domain specific use cases or like applications or high-level grammars.